File: 6_kyu/Pair_of_gloves.lua
local kata = {}

function kata.number_of_pairs(gloves)
    local count_color = {}
    for _,v in ipairs(gloves) do
        count_color[v] = 1 + (count_color[v] or 0)
    end
    local count_pairs = 0
    for _,v in pairs(count_color) do
        count_pairs = count_pairs + v // 2
    end
    return count_pairs
end

return kata

File: 6_kyu/Pair_of_gloves.go
package kata

func NumberOfPairs(gloves []string) int {
	// Prepare a map "glove color" to number of gloves
	counter := make(map[string]int, len(gloves))
	// Count the number of gloves per color
	for _, c := range gloves {
		counter[c]++
	}
	// Count the number of pairs
	numberPairs := 0
	for _, count := range counter {
		numberPairs += count / 2
	}
	// Return the number of gloves with matching color
	return numberPairs
}
